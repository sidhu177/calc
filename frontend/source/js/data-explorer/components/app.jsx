/* eslint-disable react/button-has-type, jsx-a11y/anchor-is-valid */

import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';
import classNames from 'classnames';

import {
  resetState,
  invalidateRates,
} from '../actions';

import histogramToImg from '../histogram-to-img';

import { trackEvent } from '../../common/ga';
import Description from './description';
import Highlights from './highlights';
import Histogram from './histogram';
import ProposedPrice from './proposed-price';
import ExportData from './export-data';
import ResultsTable from './results-table';
import QueryType from './query-type';
import LoadableOptionalFilters from './optional-filters/loadable-optional-filters';
import LaborCategory from './labor-category';
import LoadingIndicator from './loading-indicator';
import SearchCategory from './search-category';
import TitleTagSynchronizer from './title-tag-synchronizer';

import { autobind } from '../util';

class App extends React.Component {
  constructor(props) {
    super(props);
    autobind(this, [
      'handleSubmit',
      'handleResetClick',
      'handleDownloadClick',
    ]);
  }

  getContainerClassNames() {
    let loaded = false;
    let loading = false;
    let error = false;

    if (this.props.ratesInProgress) {
      loading = true;
    } else if (this.props.ratesError) {
      if (this.props.ratesError !== 'abort') {
        error = true;
        loaded = true;
      }
    } else {
      loaded = true;
    }

    return {
      search: true,
      content: true,
      container: true,
      loaded,
      loading,
      error,
    };
  }

  handleSubmit(e) {
    e.preventDefault();
    this.props.invalidateRates();
  }

  handleResetClick(e) {
    e.preventDefault();
    this.props.resetState();
  }

  handleDownloadClick(e) {
    e.preventDefault();
    histogramToImg(
      this.histogram.getWrappedInstance().svgEl,
      this.canvasEl,
    );
    trackEvent('download-graph', 'click');
  }

  render() {
    const prefixId = name => `${this.props.idPrefix}${name}`;

    return (
      <form
        id={prefixId('search')}
        className={classNames(this.getContainerClassNames())}
        onSubmit={this.handleSubmit}
      >
        <div className="row card dominant">
          <div className="search-header columns twelve content">
            <h2>
              Search CALC
            </h2>
            <TitleTagSynchronizer />
            <section className="search">
              <div className="container clearfix">
                <div className="row">
                  <div className="twelve columns">
                    <SearchCategory />
                    <LaborCategory api={this.props.api}>
                      <button
                        className="submit usa-button-primary icon-search"
                        aria-label="Search CALC"
                      />
                      {' '}
                      <input
                        onClick={this.handleResetClick}
                        className="reset usa-button usa-button-secondary"
                        type="reset"
                        value="Reset"
                      />
                    </LaborCategory>
                  </div>
                  <div className="twelve columns">
                    <QueryType />
                  </div>
                </div>
              </div>
            </section>
          </div>
        </div>
        <div className="row card secondary">
          <div className="columns nine">

            <div className="graph-block">
              {/* for converting the histogram into an img --> */}
              <canvas
                ref={(el) => { this.canvasEl = el; }}
                id={prefixId('graph') /* Selenium needs it. */}
                className="hidden"
                width="710"
                height="280"
              />

              <div id={prefixId('description')}>
                <Description />
              </div>

              <h4>
Hourly rate data
              </h4>

              <ProposedPrice />
              <LoadingIndicator />

              <div className="graph">
                <div id={prefixId('price-histogram')}>
                  <Histogram ref={(el) => { this.histogram = el; }} />
                </div>
              </div>

              <Highlights />

              <div className="download-buttons row">
                <div className="four columns">
                  <a
                    className="usa-button usa-button-primary"
                    id={prefixId('download-histogram') /* Selenium needs it. */}
                    href=""
                    onClick={this.handleDownloadClick}
                  >
                    â¬‡ Download graph
                  </a>
                </div>

                <div>
                  <ExportData />
                </div>

                <p className="help-text">
                  The rates shown here are fully burdened, applicable
                  {' '}
                  worldwide, and representative of the current fiscal
                  {' '}
                  year. This data represents rates awarded at the master
                  {' '}
                  contract level.
                </p>
              </div>
            </div>
          </div>

          <div className="filter-container columns three">
            <div className="filter-block">
              <h5 className="filter-title">
Optional filters
              </h5>
              <LoadableOptionalFilters />
            </div>
          </div>
        </div>
        <section className="results">
          <div className="container">
            <div className="row">
              <div className="table-container">
                <ResultsTable />
              </div>
            </div>
          </div>
        </section>
      </form>
    );
  }
}

App.propTypes = {
  api: PropTypes.object.isRequired,
  ratesInProgress: PropTypes.bool.isRequired,
  ratesError: PropTypes.string,
  resetState: PropTypes.func.isRequired,
  invalidateRates: PropTypes.func.isRequired,
  idPrefix: PropTypes.string,
};

App.defaultProps = {
  idPrefix: '',
  ratesError: null,
};

export default connect(
  state => ({
    ratesInProgress: state.rates.inProgress,
    ratesError: state.rates.error,
  }),
  { resetState, invalidateRates },
)(App);
