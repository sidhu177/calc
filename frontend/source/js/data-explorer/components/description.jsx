import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';

import DescriptionFilter from './description-filter';

import {
  MIN_EXPERIENCE,
  MAX_EXPERIENCE,
  EDU_LABELS,
  SITE_LABELS,
  BUSINESS_SIZE_LABELS,
} from '../constants';

import { scheduleLabels } from '../schedule-metadata';

import { formatCommas, stripTrailingComma } from '../util';

export function Description({
  shownResults,
  totalResults,
  minExperience,
  maxExperience,
  education,
  site,
  businessSize,
  schedule,
  laborCategory,
}) {
  let results = ' results ';
  const filtersClasses = ['filters'];
  const filters = [];

  if (laborCategory) {
    filters.push(
      <DescriptionFilter key="lab">
        {stripTrailingComma(laborCategory)}
      </DescriptionFilter>,
    );
  }

  if (education.length) {
    filters.push(
      <DescriptionFilter
        key="edu"
        label="education level"
        extraClassName="education-filter"
      >
        {education.map(x => EDU_LABELS[x]).join(', ')}
      </DescriptionFilter>,
    );
  }

  if (minExperience !== MIN_EXPERIENCE
      || maxExperience !== MAX_EXPERIENCE) {
    filters.push(
      <DescriptionFilter key="exp" label="experience">
        {minExperience}
        {' '}
-
        {maxExperience}
        {' '}
years
      </DescriptionFilter>,
    );
  }

  if (site) {
    filters.push(
      <DescriptionFilter key="sit" label="worksite">
        {SITE_LABELS[site]}
      </DescriptionFilter>,
    );
  }

  if (businessSize) {
    filters.push(
      <DescriptionFilter key="bus" label="business size">
        {BUSINESS_SIZE_LABELS[businessSize]}
      </DescriptionFilter>,
    );
  }

  if (schedule) {
    filters.push(
      <DescriptionFilter key="sch" label="schedule">
        {scheduleLabels[schedule]}
      </DescriptionFilter>,
    );
  }

  if (filters.length) {
    results += 'with ';
  } else {
    filtersClasses.push('hidden');
  }

  // TODO: The original version of this faded-in (but never out)
  // whenever it changed. We might want to do that too, or choose
  // a different animation.

  return (
    <p className="">
      {`Showing ${formatCommas(shownResults)} of `}
      <span className="total">
        {formatCommas(totalResults)}
      </span>
      {results}
      <span className={filtersClasses.join(' ')}>
        {filters}
      </span>
    </p>
  );
}

Description.propTypes = {
  shownResults: PropTypes.number.isRequired,
  totalResults: PropTypes.number.isRequired,
  minExperience: PropTypes.number.isRequired,
  maxExperience: PropTypes.number.isRequired,
  education: PropTypes.array.isRequired,
  site: PropTypes.string.isRequired,
  businessSize: PropTypes.string.isRequired,
  schedule: PropTypes.string.isRequired,
  laborCategory: PropTypes.string.isRequired,
};

function mapStateToProps(state) {
  return {
    shownResults: state.rates.data.results.length,
    totalResults: state.rates.data.count,
    minExperience: state.min_experience,
    maxExperience: state.max_experience,
    education: state.education,
    site: state.site,
    businessSize: state.business_size,
    schedule: state.schedule,
    laborCategory: state.q,
  };
}

export default connect(mapStateToProps)(Description);
