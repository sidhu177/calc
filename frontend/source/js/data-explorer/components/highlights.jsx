import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';

import {
  formatPrice,
} from '../util';

export function Highlights({
  stdDeviation,
  avgPrice,
}) {
  const stdDevMinus = avgPrice - stdDeviation;
  const stdDevPlus = avgPrice + stdDeviation;

  // TODO: The original implementation faded the proposed price in and
  // out as it was set/unset. We might want to do the same thing here.

  return (
    <div className="price-block">
      <div className="row">
        <div className="standard-deviation-block">
          <h5 className="standard-deviation-title">
Std deviation -1
          </h5>
          <h5 className="sd-highlight">
              $
            {formatPrice(stdDevMinus)}
          </h5>
        </div>
        <div className="avg-price-block">
          <h5 className="avg-price-title">
Average price
          </h5>
          <h5 className="avg-price-highlight">
              $
            {formatPrice(avgPrice)}
          </h5>
        </div>
        <div className="standard-deviation-block">
          <h5 className="standard-deviation-title">
Std deviation +1
          </h5>
          <h5 className="sd-highlight">
              $
            {formatPrice(stdDevPlus)}
          </h5>
        </div>
      </div>
    </div>
  );
}

Highlights.propTypes = {
  stdDeviation: PropTypes.number.isRequired,
  avgPrice: PropTypes.number.isRequired,
};

function mapStateToProps(state) {
  return {
    stdDeviation: state.rates.data.first_standard_deviation,
    avgPrice: state.rates.data.average,
  };
}

export default connect(mapStateToProps)(Highlights);
