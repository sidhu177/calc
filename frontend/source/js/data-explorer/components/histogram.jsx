import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';

import { select } from 'd3-selection';
import { scaleLinear } from 'd3-scale';
import { axisLeft, axisBottom } from 'd3-axis';
import { transition } from 'd3-transition';
import { extent as _extent } from 'd3-array';

import {
  templatize,
  formatCommas,
  formatPrice,
  formatFriendlyPrice,
} from '../util';

const d3 = {
  select,
  scaleLinear,
  axisLeft,
  axisBottom,
  transition,
  extent: _extent,
};

const INLINE_STYLES = `/* styles here for download graph compatibility */

* {
  vector-effect: non-scaling-stroke;
}

.stddev-rect {
  position: absolute;;
}

.bars .bar rect {
  fill: #cddc86;
}

.range-fill {
  fill: #f0f6fa;
}

.range-rule {
  stroke: #7da1b0;
  fill: none;
  stroke-dasharray: 5,5;
  stroke-width: 1;
}

.label-rule {
  stroke: #7da1b0;
}

.stddev-text {
  fill: #021014;
}

.axis .chart-label {
  fill: #436a79;
  font-style: italic;
}

.axis text {
  fill: #436a79;
  font-size: 13px;
}

.axis.x path {
  stroke: none;
}

.axis.y path {
  stroke-width: 2;
  stroke: #c5d6de;
}

.contrast-stroke {
  stroke: #61701c;
  stroke-width: 2px;
}

.tick line {
  stroke: none;
}

.stddev-text-label {
  font-size: 12px;
  fill: #436a79;
}

.avg .value,
.pp .value {
  font-weight: bold;
}

.average {
  fill: #021014;
}

.proposed {
  fill: #436a79;
}

.avg-label-box, .pp-label-box {
  fill: #fff;
  stroke: none;
}

.avg line {
  stroke-width: 1;
  stroke: #021014;
}

.pp line {
  stroke-width: 1;
  stroke: #436a79;
}`;

function updateHistogram(rootEl, data, proposedPrice, showTransition) {
  const width = 720;
  const height = 300;
  const pad = [120, 15, 60, 60];
  const top = pad[0];
  const left = pad[3];
  const right = width - pad[1];
  const bottom = height - pad[2];
  const barGap = 2;
  const svg = d3.select(rootEl)
    .attr('viewBox', [0, 0, width, height].join(' '))
    .attr('preserveAspectRatio', 'xMinYMid meet');
  const formatDollars = n => `$${formatPrice(n)}`;

  const extent = [data.minimum, data.maximum];
  const bins = data.wage_histogram;
  const x = d3.scaleLinear()
    .domain(extent)
    .range([left, right]);
  const countExtent = d3.extent(bins, d => d.count);
  const heightScale = d3.scaleLinear()
    .domain([0].concat(countExtent))
    .range([0, 1, bottom - top]);

  let stdDevMin = data.average - data.first_standard_deviation;
  let stdDevMax = data.average + data.first_standard_deviation;

  if (isNaN(stdDevMin)) stdDevMin = 0; /* eslint-disable-line no-restricted-globals */
  if (isNaN(stdDevMax)) stdDevMax = 0; /* eslint-disable-line no-restricted-globals */

  let stdDev = svg.select('.stddev');
  if (stdDev.empty()) {
    stdDev = svg.append('g')
      .attr('transform', 'translate(0,0)')
      .attr('class', 'stddev');
    stdDev.append('rect')
      .attr('class', 'range-fill');
    stdDev.append('line')
      .attr('class', 'range-rule');
    const stdDevLabels = stdDev.append('g')
      .attr('class', 'range-labels')
      .selectAll('g.chart-label')
      .data([
        { type: 'min', anchor: 'end', label: '-1 stddev' },
        { type: 'max', anchor: 'start', label: '+1 stddev' },
      ])
      .enter()
      .append('g')
      .attr('transform', 'translate(0,0)')
      .attr('class', d => `chart-label ${d.type}`);
    stdDevLabels.append('line')
      .attr('class', 'label-rule')
      .attr('y1', -5)
      .attr('y2', 5);
    const stdDevLabelsText = stdDevLabels.append('text')
      .attr('text-anchor', d => d.anchor)
      .attr('dx', (d, i) => 8 * (i ? 1 : -1));

    stdDevLabelsText.append('tspan')
      .attr('class', 'stddev-text');
    stdDevLabelsText.append('tspan')
      .attr('class', 'stddev-text-label');
  }

  let xAxis = svg.select('.axis.x');
  if (xAxis.empty()) {
    xAxis = svg.append('g')
      .attr('class', 'axis x');
  }

  let yAxis = svg.select('.axis.y');
  if (yAxis.empty()) {
    yAxis = svg.append('g')
      .attr('class', 'axis y');
  }

  let gBar = svg.select('g.bars');
  if (gBar.empty()) {
    gBar = svg.append('g')
      .attr('class', 'bars');
  }


  // draw proposed price line
  let pp = svg.select('g.pp');
  const ppOffset = -95;
  if (pp.empty()) {
    pp = svg.append('g')
      .attr('class', 'pp');

    pp.append('rect')
      .attr('y', ppOffset - 25)
      .attr('x', -55)
      .attr('class', 'pp-label-box')
      .attr('height', 26)
      .attr('width', 130)
      .attr('rx', 4)
      .attr('ry', 4);

    pp.append('text')
      .attr('text-anchor', 'middle')
      .attr('dy', ppOffset - 6)
      .attr('dx', 10)
      .attr('class', 'value proposed');

    pp.append('line');
  }

  const proposedPriceStr = formatFriendlyPrice(proposedPrice);

  // widen proposed price rect if more than a few characters long
  if (proposedPriceStr.length > 3) {
    pp.select('rect').attr('width', 150);
    pp.select('text').attr('dx', 20);
  } else {
    pp.select('rect').attr('width', 130);
    pp.select('text').attr('dx', 10);
  }

  pp.select('line')
    .attr('y1', ppOffset)
    .attr('y2', (bottom - top));
  pp.select('.value')
    .text(`$${proposedPriceStr} proposed`);

  if (proposedPrice === 0) {
    pp.style('opacity', 0);
  } else {
    pp.style('opacity', 1);
  }

  // draw average line
  let avg = svg.select('g.avg');
  const avgOffset = -55;
  if (avg.empty()) {
    avg = svg.append('g')
      .attr('class', 'avg');

    avg.append('rect')
      .attr('y', avgOffset - 23)
      .attr('x', -55)
      .attr('class', 'avg-label-box')
      .attr('width', 116)
      .attr('height', 23)
      .attr('rx', 4)
      .attr('ry', 4);

    avg.append('text')
      .attr('text-anchor', 'middle')
      .attr('dy', avgOffset - 7)
      .attr('dx', 3)
      .attr('class', 'value average');

    avg.append('line');
  }

  avg.select('line')
    .attr('y1', avgOffset)
    .attr('y2', (bottom - top));
  avg.select('.value')
    .text(`${formatDollars(data.average)} average`);

  const bars = gBar.selectAll('.bar')
    .data(bins);

  bars.exit().remove();

  const enter = bars.enter().append('g')
    .attr('class', 'bar');
  enter.append('title');

  const step = (right - left) / bins.length;
  enter.append('rect')
    .attr('x', (d, i) => left + (i * step))
    .attr('y', bottom)
    .attr('width', step)
    .attr('height', 0);
  enter.append('line')
    .attr('x1', left)
    .attr('x2', (d, i) => left + (i * step))
    .attr('y1', bottom)
    .attr('y2', bottom);

  const title = templatize('{count} results from {min} to {max}');
  bars.select('title')
    .text((d, i) => {
      const inclusive = (i === bins.length - 1);
      const sign = inclusive ? '<=' : '<';
      return title({
        count: formatCommas(d.count),
        min: formatDollars(d.min),
        sign,
        max: formatDollars(d.max),
      });
    });

  const t = showTransition
    ? d3.transition().duration(500)
    : svg;

  const stdDevWidth = x(stdDevMax) - x(stdDevMin);
  const stdDevTop = 85;
  stdDev = t.select('.stddev');
  stdDev
    .attr('transform', `translate(${[x(stdDevMin), stdDevTop]})`);

  stdDev.select('rect.range-fill')
    .attr('width', stdDevWidth)
    .attr('height', bottom - stdDevTop);

  stdDev.select('line.range-rule')
    .attr('x2', stdDevWidth);

  stdDev.select('.chart-label.min .stddev-text')
    .text(formatDollars(stdDevMin))
    .attr('x', 0)
    .attr('dy', 0);

  stdDev.select('.chart-label.min .stddev-text-label')
    .text('-1 std dev')
    .attr('x', -8)
    .attr('dy', '15px');

  stdDev.select('.chart-label.max')
    .attr('transform', `translate(${[stdDevWidth, 0]})`);

  stdDev.select('.chart-label.max .stddev-text-label')
    .text('+1 std dev')
    .attr('x', 8)
    .attr('dy', '15px');

  stdDev.select('.chart-label.max .stddev-text')
    .text(formatDollars(stdDevMax));

  const trunc = (num) => {
    if (num < 0) {
      return Math.ceil(num);
    }
    return Math.floor(num);
  };

  t.select('.avg')
    .attr('transform', `translate(${[trunc(x(data.average)), top]})`);

  t.select('.pp')
    .attr('transform', `translate(${[trunc(x(proposedPrice)), top]})`);

  t.selectAll('.bar')
    .each((d) => {
      d.x = x(d.min);
      d.width = x(d.max) - d.x;
      d.height = heightScale(d.count);
      d.y = bottom - d.height;
    })
    .select('rect')
      .attr('x', d => d.x) /* eslint-disable-line indent */
      .attr('y', d => d.y) /* eslint-disable-line indent */
      .attr('height', d => d.height) /* eslint-disable-line indent */
      .attr('width', d => d.width - barGap); /* eslint-disable-line indent */

  t.selectAll('.bar')
    .each((d) => {
      d.x = x(d.min);
      d.width = x(d.max) - d.x;
      d.height = heightScale(d.count);
      d.y = bottom - d.height;
    })
    .select('line')
      .attr('class', 'contrast-stroke') /* eslint-disable-line indent */
      .attr('x1', d => d.x) /* eslint-disable-line indent */
      .attr('x2', d => (d.x + step - barGap)) /* eslint-disable-line indent */
      .attr('y1', d => (d.y)) /* eslint-disable-line indent */
      .attr('y2', d => (d.y)); /* eslint-disable-line indent */

  const ticks = bins.map(d => d.min)
    .concat([data.maximum]);

  const xa = d3.axisBottom()
    .scale(x)
    .tickValues(ticks)
    .tickFormat((d, i) => {
      if (i === 0 || i === bins.length) {
        return formatDollars(d);
      }
      return formatPrice(d);
    });
  xAxis.call(xa)
    .attr('transform', `translate(${[0, bottom - 2]})`)
    .selectAll('.tick')
    .classed('primary', (d, i) => i === 0 || i === bins.length)
    .select('text')
    .classed('min', (d, i) => i === 0)
    .classed('max', (d, i) => i === bins.length)
    .style('text-anchor', 'end')
    .attr('transform', 'rotate(-35)');

  // remove existing labels
  svg.selectAll('text.chart-label').remove();

  xAxis.append('text')
    .attr('class', 'chart-label')
    .attr('transform', `translate(${[left + ((right - left) / 2), 45]})`)
    .attr('text-anchor', 'middle')
    .text('Ceiling price (hourly rate)');

  const yd = d3.extent(heightScale.domain());
  const ya = d3.axisLeft()
    .scale(d3.scaleLinear()
      .domain(yd)
      .range([bottom, top]))
    .tickValues(yd)
    .tickSizeOuter(0.5);
  ya.tickFormat(formatCommas);
  yAxis.call(ya)
    .attr('transform', `translate(${[left - 5, -1]})`);

  yAxis.append('text')
    .attr('class', 'chart-label')
    .attr('transform', `translate(${[-25, (height / 2) + 25]}) rotate(-90)`)
    .attr('text-anchor', 'middle')
    .text('# of results');
}

class Histogram extends React.Component {
  componentDidMount() {
    this.updateHistogram(false);
  }

  componentDidUpdate() {
    this.updateHistogram(true);
  }

  updateHistogram(showTransition) {
    updateHistogram(
      this.svgEl,
      this.props.ratesData,
      this.props.proposedPrice,
      showTransition,
    );
  }

  render() {
    return (
      <svg
        className="graph histogram has-data"
        ref={(svg) => { this.svgEl = svg; }}
      >
        <title>
Price histogram
        </title>
        <desc>
          A histogram showing the distribution of labor category prices.
          Each bar represents a range within that distribution.
        </desc>
        <style>
          {INLINE_STYLES}
        </style>
      </svg>
    );
  }
}

Histogram.propTypes = {
  ratesData: PropTypes.object.isRequired,
  proposedPrice: PropTypes.number.isRequired,
};

function mapStateToProps(state) {
  return {
    ratesData: state.rates.data,
    proposedPrice: state['proposed-price'],
  };
}

export default connect(
  mapStateToProps,
  null,
  null,
  { withRef: true },
)(Histogram);
