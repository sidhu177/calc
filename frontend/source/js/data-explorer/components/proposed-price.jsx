import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';

import { setProposedPrice } from '../actions';

export class ProposedPrice extends React.Component {
  // Note that the 'Go' button in this component does not
  // actually do anything.
  // In the legacy Data Explorer code, clicking the Go button
  // would show the proposed price on the histogram, but we now
  // do that on a per-keystroke basis.
  // We were concerned that removing the 'Go' button could make
  // things even more confusing than they already are, so for now
  // we'll do a no-op.

  constructor(props) {
    super(props);
    this.state = { typed: this.props.proposedPrice || '' };
    this.handleChange = this.handleChange.bind(this);
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.proposedPrice !== this.props.proposedPrice) {
      const typedFloat = parseFloat(this.state.typed || '0');
      if (typedFloat !== nextProps.proposedPrice) {
        this.setState({ typed: nextProps.proposedPrice || '' });
      }
    }
  }

  componentDidUpdate(_, prevState) {
    if (prevState.typed !== this.state.typed) {
      const value = this.state.typed;
      const floatValue = value ? parseFloat(value) : 0;
      if (!isNaN(floatValue) && floatValue >= 0) { /* eslint-disable-line no-restricted-globals */
        this.props.setProposedPrice(floatValue);
      }
    }
  }

  handleChange(e) {
    const { value } = e.target;
    const filteredChars = [];
    let decimalFound = false;
    let centDigits = 0;

    for (let i = 0; i < value.length; i++) {
      const c = value[i];

      if (/[0-9]/.test(c) && centDigits < 2) {
        filteredChars.push(c);
        if (decimalFound) {
          centDigits++;
        }
      } else if (c === '.' && !decimalFound) {
        filteredChars.push(c);
        decimalFound = true;
      }
    }

    this.setState({ typed: filteredChars.join('') });
  }

  render() {
    // Note that Chrome has issues with number input, so we're just
    // going to use a standard text field here. This is unfortunate for
    // mobile users, since they won't get a numeric keypad.
    //
    // See more details at: https://github.com/facebook/react/issues/1549

    const id = `${this.props.idPrefix}proposed-price`;

    return (
      <div className="proposed-price-block">
        <h5 className="proposed-price-title">
          Proposed price
        </h5>
        <div className="proposed-price">
          <label htmlFor={id} className="usa-sr-only">
            Proposed price
          </label>
          <input
            id={id}
            type="text"
            name="proposed-price"
            className="form__inline"
            placeholder="$"
            value={this.state.typed}
            onChange={this.handleChange}
          />
          <button
            type="button"
            className="button-primary go"
          >
            Go
          </button>
        </div>
      </div>
    );
  }
}

ProposedPrice.propTypes = {
  proposedPrice: PropTypes.number.isRequired,
  setProposedPrice: PropTypes.func.isRequired,
  idPrefix: PropTypes.string,
};

ProposedPrice.defaultProps = {
  idPrefix: '',
};

export default connect(
  state => ({ proposedPrice: state['proposed-price'] }),
  { setProposedPrice },
)(ProposedPrice);
