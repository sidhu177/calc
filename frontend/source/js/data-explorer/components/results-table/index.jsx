import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';

import { setSort } from '../../actions';
import createSortableColumn from './sortable-column';
import * as ExcludedColumn from './excluded-column';
import * as LaborCategoryColumn from './labor-category-column';
import * as EducationColumn from './education-column';
import * as ExperienceColumn from './experience-column';
import * as PriceColumn from './price-column';
import * as VendorColumn from './vendor';

const COLUMNS = [
  ExcludedColumn,
  LaborCategoryColumn,
  PriceColumn,
  EducationColumn,
  ExperienceColumn,
  VendorColumn,
  createSortableColumn({
    key: 'schedule',
    title: 'Contract vehicle',
  }),
];

const { priceForContractYear } = PriceColumn;

export class ResultsTable extends React.Component {
  renderBodyRows() {
    return this.props.results
      .filter(r => !!priceForContractYear(this.props.contractYear, r))
      .map(result => (
        <tr key={result.id}>
          {COLUMNS.map((col) => {
            const cellKey = `${result.id}-${col.DataCell.cellKey}`;
            return (
              <col.DataCell key={cellKey} sort={this.props.sort} result={result} />
            );
          })}
        </tr>
      ));
  }

  render() {
    const id = `${this.props.idPrefix}results-table`;
    const idHref = `#${id}`;

    return (
      <table id={id} className="results has-data sortable hoverable">
        <thead>
          <tr>
            {COLUMNS.map(col => (
              <col.HeaderCell
                key={`header-${col.DataCell.cellKey}`}
                setSort={this.props.setSort}
                sort={this.props.sort}
              />
            ))}
          </tr>
        </thead>
        <tbody>
          {this.renderBodyRows()}
        </tbody>
      </table>
    );
  }
}

ResultsTable.propTypes = {
  sort: PropTypes.object.isRequired,
  setSort: PropTypes.func.isRequired,
  results: PropTypes.array.isRequired,
  contractYear: PropTypes.string.isRequired,
  idPrefix: PropTypes.string,
};

ResultsTable.defaultProps = {
  idPrefix: '',
};

function mapStateToProps(state) {
  return {
    sort: state.sort,
    results: state.rates.data.results,
    contractYear: state['contract-year'],
  };
}

const mapDispatchToProps = { setSort };

export default connect(mapStateToProps, mapDispatchToProps)(ResultsTable);
