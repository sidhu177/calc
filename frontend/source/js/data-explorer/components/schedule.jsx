import PropTypes from 'prop-types';
import React from 'react';
import { connect } from 'react-redux';

import {
  setSchedule as setScheduleAction,
  setQueryBy as setQueryByAction
} from '../actions';
import { QUERY_BY_SCHEDULE } from '../constants';
import { scheduleLabels } from '../schedule-metadata';

export function Schedule({
  selectedSchedule, setSchedule, setQueryBy, queryBy
}) {
  const handleChange = (e) => {
    setSchedule(e.target.value);
    setQueryBy(QUERY_BY_SCHEDULE);
  };
  const defaultMsg = `In all ${Object.keys(scheduleLabels).length} contract vehicles`;
  // In most instances, we display legacy schedules as "Legacy Schedule," i.e., "Legacy MOBIS."
  // Here, however, we want to display the "Legacy" modifier in parenthesis after the name.
  // Since the legacy modifier only is found in schedule.full_name, we have to regex.
  const legacyPrefix = "Legacy ";
  const makeInput = (value, label) => {
    const id = value.replace(/ /g, '-').toLowerCase() || 'all-schedules';
    const makeLabel = (title) => {
      let scheduleLabel = title;
      let labelSuffix;
      if (title.includes(legacyPrefix)) {
        scheduleLabel = title.replace(legacyPrefix, '');
        labelSuffix = "(Legacy)";
      }
      return { scheduleLabel, labelSuffix };
    };
    const { scheduleLabel, labelSuffix } = makeLabel(label);
    return (
      <li key={id}>
        <input
          type="radio"
          id={id}
          name={id}
          value={value}
          onChange={handleChange}
          checked={selectedSchedule === value && queryBy === QUERY_BY_SCHEDULE}
        />
        <label htmlFor={id}>
          {scheduleLabel}
          <span className="label__suffix">
            {labelSuffix}
          </span>
        </label>
      </li>
    );
  };

  const makeChoices = labels => [
    { key: '', value: '', label: defaultMsg },
  ].concat(Object.keys(labels).map(
    value => ({ value, label: labels[value] }),
  )).map(({ value, label }) => (
    makeInput(value, label)
  ));

  return (
    <ul className="filter--schedule">
      {makeChoices(scheduleLabels, defaultMsg)}
    </ul>
  );
}

Schedule.propTypes = {
  selectedSchedule: PropTypes.string.isRequired,
  setSchedule: PropTypes.func.isRequired,
  setQueryBy: PropTypes.func.isRequired,
  queryBy: PropTypes.string.isRequired,
};

export default connect(
  state => ({
    selectedSchedule: state.schedule,
    queryBy: state.query_by,
  }),
  {
    setSchedule: setScheduleAction,
    setQueryBy: setQueryByAction,
  },
)(Schedule);
