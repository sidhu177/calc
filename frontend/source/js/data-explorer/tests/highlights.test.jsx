import toJson from 'enzyme-to-json';

import { Highlights } from '../components/highlights';
import makeSetup from './testSetup';

const defaultProps = {
  stdDeviation: 1.1,
  avgPrice: 2.1,
  proposedPrice: 5,
};

const setup = makeSetup(Highlights, defaultProps);

describe('<Highlights>', () => {
  it('renders correctly', () => {
    const { wrapper } = setup();

    expect(wrapper.find('.sd-highlight').first().text()).toBe('$1');
    expect(wrapper.find('.avg-price-highlight').text()).toBe('$2');
    expect(wrapper.find('.sd-highlight').last().text()).toBe('$3');
    expect(wrapper.find('.proposed-price-highlight').text()).toBe('$5');
  });

  it('matches snapshot', () => {
    const { wrapper } = setup();
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
