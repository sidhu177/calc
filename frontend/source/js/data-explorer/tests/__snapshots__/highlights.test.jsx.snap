// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Highlights> matches snapshot 1`] = `
<div
  className="highlights-container row"
>
  <div
    className="price-block"
  >
    <div
      className="row"
    >
      <div
        className="standard-deviation-block"
      >
        <h5
          className="standard-deviation-title"
        >
          Std deviation -1
        </h5>
        <h5
          className="sd-highlight"
        >
          $
          1
        </h5>
      </div>
      <div
        className="avg-price-block"
      >
        <h5
          className="avg-price-title"
        >
          Average
        </h5>
        <h5
          className="avg-price-highlight"
        >
          $
          2
        </h5>
      </div>
      <div
        className="standard-deviation-block"
      >
        <h5
          className="standard-deviation-title"
        >
          Std deviation +1
        </h5>
        <h5
          className="sd-highlight"
        >
          $
          3
        </h5>
      </div>
    </div>
  </div>
  <div
    className="proposed-price-block"
    style={
      Object {
        "display": "block",
      }
    }
  >
    <h5
      className="proposed-price-title"
    >
      Proposed price
    </h5>
    <h5
      className="proposed-price-highlight"
    >
      $
      5
    </h5>
  </div>
</div>
`;
